zarr/indexing.py

```diff
@@ -0,0 +1,676 @@
+        else:
+            self.drop_axes = None
+
+    def __iter__(self):
+        for dim_projections in itertools.product(*self.dim_indexers):
+
+            chunk_coords = tuple(p.dim_chunk_ix for p in dim_projections)
+            chunk_selection = tuple(p.dim_chunk_sel for p in dim_projections)
+            out_selection = tuple(p.dim_out_sel for p in dim_projections if p.dim_out_sel is not None)
+
+            # handle advanced indexing arrays orthogonally
+            if self.is_advanced:
+                # numpy doesn't support orthogonal indexing directly as yet, so need to work
+                # around via np.ix_. Also np.ix_ does not support a mixture of arrays and slices
+                # or integers, so need to convert slices and integers into ranges.
+                chunk_selection = ix_(*chunk_selection)
+
+                # special case for non-monotonic indices
+                if any([not isinstance(s, (int, slice)) for s in out_selection]):
+                    out_selection = ix_(*out_selection)
+
+            yield ChunkProjection(chunk_coords, chunk_selection, out_selection)
+
+
+class OIndex(object):
+
+    def __init__(self, array):
+        self.array = array
+
+    def __getitem__(self, selection):
+        return self.array.get_orthogonal_selection(selection)
+
+    def __setitem__(self, selection, value):
+        return self.array.set_orthogonal_selection(selection, value)
+
+
+def is_coordinate_selection(selection, array):
+    return (
+        (len(selection) == len(array._shape)) and
+        all(
+            [is_integer(dim_sel) or is_integer_array(dim_sel)
+             for dim_sel in selection]
+        )
+    )
+
+
+def is_mask_selection(selection, array):
+    return (
+        len(selection) == 1 and
+        is_bool_array(selection[0]) and
+        selection[0].shape == array.shape
+    )
+
+
+def replace_lists(selection):
+    return tuple(
+        np.asarray(dim_sel) if isinstance(dim_sel, list) else dim_sel
+        for dim_sel in selection
+    )
+
+
+# noinspection PyProtectedMember
+class CoordinateIndexer(object):
+
+    def __init__(self, selection, array):
+
+        # some initial normalization
+        selection = ensure_tuple(selection)
+        selection = tuple([i] if is_integer(i) else i for i in selection)
+        selection = replace_lists(selection)
+
+        # validation
+        if not is_coordinate_selection(selection, array):
+            # TODO refactor error messages for consistency
+            raise IndexError('invalid coordinate selection')
+
+        # attempt to broadcast selection - this will raise error if array dimensions don't match
+        selection = np.broadcast_arrays(*selection)
+
+        # normalization
+        for dim_sel, dim_len in zip(selection, array.shape):
+
+            # check number of dimensions, only support indexing with 1d array
+            if len(dim_sel.shape) > 1:
```

Comment:
I'm not sure I'm reading this right, but does this mean you only support vectorized indexing with 1D arrays?

Vectorized indexing with >1D arrays should be pretty easy and can be quite useful. You just need to flatten the indices after broadcasting and unflatten the result.

Reply:
Thank you for this tip, I finally get coordinate indexing (at least without slices)! I've added support for multi-dimensional coordinate arrays.